src/vidpf.rs

```diff
@@ -0,0 +1,822 @@
+            let mut state_1 = VidpfEvalState::init_from_key(key_1);
+
+            let n = input.len();
+            for level in 0..n {
+                let share_0;
+                let share_1;
+                (state_0, share_0) = vidpf
+                    .eval_next(key_0.id, public, input, level, &state_0, nonce)
+                    .unwrap();
+                (state_1, share_1) = vidpf
+                    .eval_next(key_1.id, public, input, level, &state_1, nonce)
+                    .unwrap();
+
+                assert_eq!(
+                    share_0 + share_1,
+                    *weight,
+                    "shares must add up to the expected weight at the current level: {:?}",
+                    level
+                );
+
+                assert_eq!(
+                    state_0.proof, state_1.proof,
+                    "proofs must be equal at the current level: {:?}",
+                    level
+                );
+            }
+        }
+    }
+
+    mod weight {
+        use std::io::Cursor;
+        use subtle::{Choice, ConditionallyNegatable};
+
+        use crate::{
+            codec::{Encode, ParameterizedDecode},
+            idpf::IdpfValue,
+            vdaf::xof::{Seed, Xof, XofTurboShake128},
+        };
+
+        use super::{TestWeight, TEST_WEIGHT_LEN};
+
+        #[test]
+        fn roundtrip_codec() {
+            let weight = TestWeight::new(&[21.into(), 22.into(), 23.into()]).unwrap();
+
+            let mut bytes = vec![];
+            weight.encode(&mut bytes).unwrap();
+
+            let expected_bytes = [
+                [vec![21], vec![0u8; 15]].concat(),
+                [vec![22], vec![0u8; 15]].concat(),
+                [vec![23], vec![0u8; 15]].concat(),
+            ]
+            .concat();
+
+            assert_eq!(weight.encoded_len().unwrap(), expected_bytes.len());
+            // Check endianness of encoding
+            assert_eq!(bytes, expected_bytes);
+
+            let decoded =
+                TestWeight::decode_with_param(&TEST_WEIGHT_LEN, &mut Cursor::new(&bytes)).unwrap();
+            assert_eq!(weight, decoded);
+        }
+
+        #[test]
+        fn add_sub() {
+            let [a, b] = compatible_weights();
+            let mut c = a.clone();
+            c += a.clone();
+
+            assert_eq!(
+                (a.clone() + b.clone()) + (a.clone() - b.clone()),
+                c,
+                "a: {:?} b:{:?}",
+                a,
+                b
+            );
+        }
+
+        #[test]
+        fn conditional_negate() {
+            let [a, _] = compatible_weights();
+            let mut c = a.clone();
+            c.conditional_negate(Choice::from(0));
+            let mut d = a.clone();
+            d.conditional_negate(Choice::from(1));
+            let zero = TestWeight::zero(&TEST_WEIGHT_LEN);
+
+            assert_eq!(c + d, zero, "a: {:?}", a);
+        }
+
+        #[test]
+        #[should_panic]
```

Comment:
We can tighten these tests by matching against the panic message.
```suggestion
        #[should_panic = "invalid weight length"]
```